src/types.rs

```diff
@@ -121,6 +131,77 @@ impl ResponseParam {
         }
     }
 }
+#[derive(Deserialize, Debug, Serialize)]
+pub struct EventRequest {
+    pub tx_hash: DisplayBytes,
+    pub network: String,
+}
+
+#[derive(Deserialize, Debug, Serialize)]
+pub struct TxLog {
+    pub address: DisplayBytes,
+    pub data: DisplayBytes,
+    pub index: String,
+    pub topics: [Option<Hash>; 4],
+}
+
+#[derive(Deserialize, Debug, Serialize)]
+pub struct EventResponse {
+    pub events: Vec<DecodedEvent>,
+}
+
+impl EventResponse {
+    pub fn new(events: Vec<(Event, Log, String)>) -> EventResponse {
+        EventResponse {
+            events: events
+                .iter()
+                .map(|(e, l, i)| DecodedEvent::new(e, l, i))
+                .collect(),
+        }
+    }
+}
+
+#[derive(Deserialize, Debug, Serialize)]
+pub struct DecodedEvent {
+    pub name: String,
+    pub inputs: Vec<DecodedEventParam>,
+    pub index: String,
+}
+
+impl DecodedEvent {
+    pub fn new(event: &Event, log: &Log, index: &str) -> DecodedEvent {
```

Comment:
if you need something by value (aka "cloned") inside your function, then accept it by value!
look at this two code examples
```rust
fn new(some: &Value) -> Foo {
  Foo { some: some.clone() }
}

fn main() {
  // if we need the value after
  let some = Value::new();
  foo(&some); // clone inside
  
  // if we don't need the value
  foo(&some); // we can't say "hey dude don't clone this I don't need it"
  // some is dropped here
}
// total clones: 2
```

```rust
fn new(some: Value) -> Foo {
  Foo { some: some }
}

fn main() {
  // if we need the value after
  let some = Value::new();
  foo(some.clone()); // we can clone when we're calling
  
  // if we don't need the value
  foo(some); // we can move
}
// total clones: 1
```

As you can see, the second variant just always better, it will have less clones.
You can accept by reference, but only if you'll use it by reference, for example
```rust
fn foo(some: &Value) {
  println!("{}, some); // we use by reference 
}
```
we don't need to clone inside, so reference is okay